components/places/ios/Places/Places.swift

```diff
@@ -0,0 +1,418 @@
+        }
+    }
+
+    deinit {
+        // Note: we shouldn't need to queue.sync with our queue in deinit (no more references
+        // exist to us), however we still need to sync with the write conn's queue, since it
+        // could still be in use.
+
+        self.writeConn.queue.sync {
+            // If the writer is still around (it should be), return it to the api.
+            let writeHandle = self.writeConn.takeHandle()
+            if writeHandle != 0 {
+                PlacesError.unwrapOrLog { error in
+                    places_api_return_write_conn(self.handle, writeHandle, error)
+                }
+            }
+        }
+
+        PlacesError.unwrapOrLog { error in
+            places_api_destroy(self.handle, error)
+        }
+    }
+
+    /**
+     * Open a new reader connection.
+     *
+     * - Throws: `PlacesError` if a connection could not be opened.
+     */
+    func openReader() throws -> PlacesReadConn {
+        return try queue.sync {
+            let h = try PlacesError.unwrap { error in
+                places_connection_new(handle, Int32(PlacesConn_ReadOnly), error)
+            }
+            return PlacesReadConn(handle: h, api: self)
+        }
+    }
+
+    /**
+     * Get the writer connection.
+     *
+     * - Note: There is only ever a single writer connection,
+     *         and it's opened when the database is constructed,
+     *         so this function does not throw
+     */
+    func getWriter() -> PlacesWriteConn {
+        return queue.sync {
+            self.writeConn
+        }
+    }
+}
+
+/**
+ * A read-only connection to the places database.
+ */
+public class PlacesReadConn {
+    fileprivate let queue = DispatchQueue(label: "com.mozilla.places.conn")
+    fileprivate var handle: ConnHandle;
+    fileprivate weak var api: PlacesApi?
+
+    fileprivate init(handle: ConnHandle, api: PlacesApi? = nil) {
+        self.handle = handle
+        self.api = api
+    }
+
+    // Note: caller synchronizes!
+    fileprivate func checkApi() throws {
+        if api == nil {
+            throw PlacesError.connUseAfterApiClosed
+        }
+    }
+
+    // Note: caller synchronizes!
+    fileprivate func takeHandle() -> ConnHandle {
+        let handle = self.handle
+        self.handle = 0
+        return handle
+    }
+
+    deinit {
+        // Note: don't need to queue.sync in deinit -- no more references exist to us.
+        let handle = self.takeHandle()
+        if handle != 0 {
+            // In practice this can only fail if the rust code panics, which for this
+            // function would be quite bad.
+            try! PlacesError.tryUnwrap({ err in
+                places_connection_destroy(handle, err)
+            })
+        }
+    }
+
+    /**
+     * Returns the bookmark subtree rooted at `rootGUID`.
+     *
+     * This differs from `getBookmark` in that it populates folder children
+     * recursively (specifically, any `BookmarkFolder`s in the returned value
```

Comment:
Should this method have an optional boolean with a default to indicate whether or not to get the tree recursively? Just thinking ahead to usage, it is likely we only want to fetch one level of the tree at a time. So, it would be nice if we could pass `recursive: false` to this.

Reply:
This is the distinction between getBookmark and getBookmarkTree, but if you'd prefer they be one method with a boolean that's doable.

Reply:
That's not how I understood it. According to the comments on `getBookmark`, it never populates the `children`, only the `childGUIDs`, so it wouldn't even be able to return a single level of bookmarks, only their GUIDs.

To be clear, I still see a need for `getBookmark()`, for when you only have a GUID and want to populate the rest of the details for it aside from `children`. But for `getBookmarksTree()`, I think it is currently doing the right thing (populating the `children` bookmarks), but the flag I'm requesting would be to prevent populating the children of those children.

Reply:
Added in https://github.com/mozilla/application-services/pull/743/commits/be14d85e1dc99442de7d8f1764c5e88ed5d50aa2